{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f5f701f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy\n",
    "import hdf5plugin\n",
    "import pandas\n",
    "import numpy\n",
    "import anndata\n",
    "import scipy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "42caef52",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_names = [\n",
    "    \"MEF\",\n",
    "    \"Pool1_mESC\",\n",
    "    \"Pool2_mESC\",\n",
    "    \"Repro_Day2_BFP\",\n",
    "    \"Repro_Day2_Hic2\",\n",
    "    \"Repro_Day4_BFP\",\n",
    "    \"Repro_Day4_Hic2\",\n",
    "    \"Repro_Day6_BFP\",\n",
    "    \"Repro_Day6_Hic2\",\n",
    "    \"Repro_Day9_BFP\",\n",
    "    \"Repro_Day9_Hic2\",\n",
    "    \"Repro_Day12_BFP\",\n",
    "    \"Repro_Day12_Hic2\",\n",
    "]\n",
    "\n",
    "seurat_batches = [\n",
    "    'MEF',\n",
    "    'Pool1_mESC',\n",
    "    'Pool2_mESC',\n",
    "    'Repro_Day2_BFP',\n",
    "    'Repro_Day2_Hic2',\n",
    "    'Repro_Day4_BFP',\n",
    "    'Repro_Day4_Hic2',\n",
    "    'Repro_Day6_BFP',\n",
    "    'Repro_Day6_Hic2',\n",
    "    'Repro_Day9_BFP',\n",
    "    'Repro_Day9_Hic2',\n",
    "    'Repro_Day12_BFP',\n",
    "    'Repro_Day12_Hic2',\n",
    "]\n",
    "\n",
    "\n",
    "groups = [\n",
    "    \"MEF\",\n",
    "    \"mESC\",\n",
    "    \"mESC\",\n",
    "    \"control\",\n",
    "    \"Hic2\",\n",
    "    \"control\",\n",
    "    \"Hic2\",\n",
    "    \"control\",\n",
    "    \"Hic2\",\n",
    "    \"control\",\n",
    "    \"Hic2\",\n",
    "    \"control\",\n",
    "    \"Hic2\",\n",
    "]\n",
    "\n",
    "days = [\n",
    "    0,\n",
    "    0,\n",
    "    0,\n",
    "    2,\n",
    "    2,\n",
    "    4,\n",
    "    4,\n",
    "    6,\n",
    "    6,\n",
    "    9,\n",
    "    9,\n",
    "    12,\n",
    "    12,\n",
    "]\n",
    "\n",
    "samples = [\n",
    "    \"MEF\",\n",
    "    \"mESC\",\n",
    "    \"mESC\",\n",
    "    \"Day 2 Control\",\n",
    "    \"Day 2 Hic2 OX\",\n",
    "    \"Day 4 Control\",\n",
    "    \"Day 4 Hic2 OX\",\n",
    "    \"Day 6 Control\",\n",
    "    \"Day 6 Hic2 OX\",\n",
    "    \"Day 9 Control\",\n",
    "    \"Day 9 Hic2 OX\",\n",
    "    \"Day 12 Control\",\n",
    "    \"Day 12 Hic2 OX\"\n",
    "]\n",
    "\n",
    "sample_order = [\n",
    "    \"MEF\",\n",
    "    \"Day 2 Control\",\n",
    "    \"Day 4 Control\",\n",
    "    \"Day 6 Control\",\n",
    "    \"Day 9 Control\",\n",
    "    \"Day 12 Control\",\n",
    "    \"Day 2 Hic2 OX\",\n",
    "    \"Day 4 Hic2 OX\",\n",
    "    \"Day 6 Hic2 OX\",\n",
    "    \"Day 9 Hic2 OX\",\n",
    "    \"Day 12 Hic2 OX\",\n",
    "    \"mESC\",\n",
    "]\n",
    "\n",
    "sample_colors = [\n",
    "    '#cbc9ca',\n",
    "    \"#ffd1be\", \n",
    "    '#f3ad91',\n",
    "    '#eb684b',\n",
    "    '#da2e24',\n",
    "    '#a81918',\n",
    "    \"#ced8f0\",\n",
    "    '#b6d8e5',\n",
    "    '#71add9',\n",
    "    '#3180bd',\n",
    "    '#164f9b',\n",
    "    '#0a652c'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "166d0f51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read Seurat object to add UMAP and FDL\n",
    "adata_seurat = scanpy.read_h5ad(\"/scratch/lema/m26_losu/seurat_object.h5ad\")\n",
    "fdl = pandas.read_csv(\"/home/avesta/m26_losu/Hic2/fdl.csv\", index_col=0)\n",
    "adata_seurat.obsm['X_fdl'] = numpy.array(fdl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e46d5f9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to convert mtx to anndata\n",
    "\n",
    "def mtx_to_anndata(path, filter_cells=True):\n",
    "    \"\"\"Generate an anndata object from the STAR aligner output folder.\n",
    "    Based on https://github.com/alexdobin/STAR/issues/774#issuecomment-850477636.\"\"\"\n",
    "    path=path\n",
    "    # Load Read Counts\n",
    "    X = scanpy.read_mtx(path+'Velocyto/raw/spliced.mtx')\n",
    "\n",
    "    # Transpose counts matrix to have Cells as rows and Genes as cols as expected by AnnData objects\n",
    "    X = X.X.transpose()\n",
    "\n",
    "    # Load the 3 matrices containing Spliced, Unspliced and Ambigous reads\n",
    "    mtxU = numpy.loadtxt(path+'Velocyto/raw/unspliced.mtx', skiprows=3, delimiter=' ')\n",
    "    mtxS = numpy.loadtxt(path+'Velocyto/raw/spliced.mtx', skiprows=3, delimiter=' ')\n",
    "    mtxA = numpy.loadtxt(path+'Velocyto/raw/ambiguous.mtx', skiprows=3, delimiter=' ')\n",
    "    mtxG = numpy.loadtxt(path+'Gene/raw/matrix.mtx', skiprows=3, delimiter=' ')\n",
    "\n",
    "    # Extract sparse matrix shape informations from the third row\n",
    "    shapeU = numpy.loadtxt(path+'Velocyto/raw/unspliced.mtx', skiprows=2, max_rows = 1 ,delimiter=' ')[0:2].astype(int)\n",
    "    shapeS = numpy.loadtxt(path+'Velocyto/raw/spliced.mtx', skiprows=2, max_rows = 1 ,delimiter=' ')[0:2].astype(int)\n",
    "    shapeA = numpy.loadtxt(path+'Velocyto/raw/ambiguous.mtx', skiprows=2, max_rows = 1 ,delimiter=' ')[0:2].astype(int)\n",
    "    shapeG = numpy.loadtxt(path+'Gene/raw/matrix.mtx', skiprows=2, max_rows = 1 ,delimiter=' ')[0:2].astype(int)\n",
    "\n",
    "    # Read the sparse matrix with csr_matrix((data, (row_ind, col_ind)), shape=(M, N))\n",
    "    # Subract -1 to rows and cols index because csr_matrix expects a 0 based index\n",
    "    # Traspose counts matrix to have Cells as rows and Genes as cols as expected by AnnData objects\n",
    "\n",
    "    spliced = scipy.sparse.csr_matrix((mtxS[:,2], (mtxS[:,0]-1, mtxS[:,1]-1)), shape = shapeS).transpose()\n",
    "    unspliced = scipy.sparse.csr_matrix((mtxU[:,2], (mtxU[:,0]-1, mtxU[:,1]-1)), shape = shapeU).transpose()\n",
    "    ambiguous = scipy.sparse.csr_matrix((mtxA[:,2], (mtxA[:,0]-1, mtxA[:,1]-1)), shape = shapeA).transpose()\n",
    "    any = scipy.sparse.csr_matrix((mtxG[:,2], (mtxG[:,0]-1, mtxG[:,1]-1)), shape = shapeG).transpose()\n",
    "\n",
    "    # Load Genes and Cells identifiers\n",
    "    obs = pandas.read_csv(path+'Velocyto/raw/barcodes.tsv',\n",
    "                  header = None, index_col = 0)\n",
    "\n",
    "    # Remove index column name to make it compliant with the anndata format\n",
    "    obs.index.name = None\n",
    "    obs[\"barcode\"] = obs.index\n",
    "\n",
    "    var = pandas.read_csv(path+'Velocyto/raw/features.tsv', sep='\\t', usecols=[0, 1], header=None,\n",
    "                                    names = [\"gene_ids\", \"gene_names\"], index_col = 0)\n",
    "  \n",
    "    # Build AnnData object to be used with ScanPy and ScVelo\n",
    "    adata = anndata.AnnData(X = X, obs = obs, var = var,\n",
    "                                                 layers = {'spliced': spliced, 'unspliced': unspliced,\n",
    "                                                           'ambiguous': ambiguous, \"matrix\": any})\n",
    "    adata.var_names_make_unique()\n",
    "\n",
    "    # Subset Cells based on STAR filtering\n",
    "    if filter_cells:\n",
    "        selected_barcodes = pandas.read_csv(path+'Gene/filtered/barcodes.tsv', header = None)\n",
    "        adata = adata[selected_barcodes[0]]\n",
    "\n",
    "    return adata.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f950eca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: MEF\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Pool1_mESC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Pool2_mESC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Repro_Day2_BFP\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Repro_Day2_Hic2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Repro_Day4_BFP\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Repro_Day4_Hic2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Repro_Day6_BFP\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Repro_Day6_Hic2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Repro_Day9_BFP\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Repro_Day9_Hic2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Repro_Day12_BFP\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading file: Repro_Day12_Hic2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2288664/2761841717.py:37: ImplicitModificationWarning: Setting element `.obsm['X_umap']` of view, initializing view as actual.\n",
      "  adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n",
      "/home/avesta/m26_losu/miniconda3/envs/velocity/lib/python3.12/site-packages/anndata/_core/anndata.py:1774: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "# Read and combined mtx files\n",
    "\n",
    "adata_files = []\n",
    "directory_path = \"/scratch/lema/m26_losu/star_counts_gencode\"\n",
    "\n",
    "for index, base_name in enumerate(base_names):\n",
    "    star_path = f\"{directory_path}/{base_name}/\"\n",
    "    group = groups[index]\n",
    "    day = days[index]\n",
    "    print(f\"Loading file: {base_name}\")\n",
    "    \n",
    "    # Read loom file\n",
    "    adata_star = mtx_to_anndata(star_path, filter_cells=False)\n",
    "    adata_star.obs[\"batch\"] = base_name\n",
    "    adata_star.obs[\"sample\"] = samples[index]\n",
    "    adata_star.obs[\"group\"] = groups[index]\n",
    "    adata_star.obs[\"day\"] = days[index]\n",
    "    adata_star.var_names_make_unique()\n",
    "\n",
    "    # Add barcode variable\n",
    "    adata_star.obs[\"barcode\"] = adata_star.obs.index\n",
    "    adata_star.strings_to_categoricals()\n",
    "\n",
    "    # Add UMAP and FDL from Seurat object\n",
    "    seurat_batch = seurat_batches[index]\n",
    "    sample_seurat = adata_seurat[adata_seurat.obs[\"sample\"]==seurat_batch]\n",
    "    barcodes_seurat = set(sample_seurat.obs[\"barcode\"])\n",
    "\n",
    "    # Filter out barcodes not in Seurat object\n",
    "    adata_star = adata_star[adata_star.obs[\"barcode\"].isin(barcodes_seurat)]\n",
    "\n",
    "    # Add umap from Seurat object\n",
    "    umap_map = pandas.DataFrame(\n",
    "        sample_seurat.obsm['X_umap'], \n",
    "        index=sample_seurat.obs['barcode']\n",
    "    )\n",
    "    adata_star.obsm['X_umap'] = umap_map.loc[adata_star.obs[\"barcode\"]].values\n",
    "\n",
    "    # Add FDL\n",
    "    fdl_map = pandas.DataFrame(\n",
    "        sample_seurat.obsm['X_fdl'], \n",
    "        index=sample_seurat.obs['barcode']\n",
    "    )\n",
    "    adata_star.obsm['X_fdl'] = fdl_map.loc[adata_star.obs[\"barcode\"]].values\n",
    "\n",
    "    adata_files.append(adata_star)\n",
    "\n",
    "# Concatenate Anndata objects\n",
    "\n",
    "adata = adata_files[0].concatenate(\n",
    "    adata_files[1:],\n",
    "    join=\"outer\",\n",
    "    batch_key=\"batch\",\n",
    "    batch_categories=base_names,\n",
    "    index_unique=\"-\"\n",
    ")\n",
    "\n",
    "# Assign colors to samples\n",
    "adata.obs[\"sample\"] = adata.obs[\"sample\"].cat.reorder_categories(sample_order)\n",
    "adata.uns[\"sample_colors\"] = sample_colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0a75d718",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 13315 × 78334\n",
      "    obs: 'barcode', 'batch', 'sample', 'group', 'day'\n",
      "    var: 'gene_names'\n",
      "    uns: 'sample_colors'\n",
      "    obsm: 'X_umap', 'X_fdl'\n",
      "    layers: 'spliced', 'unspliced', 'ambiguous', 'matrix'\n"
     ]
    }
   ],
   "source": [
    "print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "92ed0911",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Choose gene names as variable names\n",
    "\n",
    "adata.var[\"ensemble_ids\"] = adata.var_names\n",
    "adata.var_names = adata.var[\"gene_names\"].astype(str)\n",
    "if not adata.var_names.is_unique:\n",
    "    # Make duplicates unique by adding suffix\n",
    "    adata.var_names_make_unique()\n",
    "# Rename gene name\n",
    "adata.var['gene_symbol'] = adata.var['gene_names']\n",
    "adata.var = adata.var.drop(columns='gene_names')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "009bb06d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make observation names unique\n",
    "adata.obs_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a5ca57f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 13315 × 78334\n",
      "    obs: 'barcode', 'batch', 'sample', 'group', 'day'\n",
      "    var: 'ensemble_ids', 'gene_symbol'\n",
      "    uns: 'sample_colors'\n",
      "    obsm: 'X_umap', 'X_fdl'\n",
      "    layers: 'spliced', 'unspliced', 'ambiguous', 'matrix'\n"
     ]
    }
   ],
   "source": [
    "print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1694396b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(\n",
    "    \"/scratch/lema/m26_losu/splice_counts_star_fdl.h5ad\",\n",
    "    compression=hdf5plugin.FILTERS[\"zstd\"]\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "velocity",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
